README.rst
===================
lcheapo_noobspy
===================

Python tools for processing LCHEAPO 2000 data files

All tools except `lcdump` and `lcinfo` create/append to a process-steps file
according to the SDPCHAIN rules

If you want to plot the data, or convert to SeisCompData structure miniSEED files, go to lcheapo_obspy_

.. _lcheapo_obspy: https://github.com/WayneCrawford/lcheapo_obspy

Overview
======================

Command-line programs:
----------------------

:lcdump: dump raw information from LCHEAPO files
:lcinfo: return basic information about an LCHEAPO file
:lccut: extract section of an LCHEAPO file
:lcfix: fix common bugs in an LCHEAPO file
:lcheader: create an LCHEAPO header + directory
:sdpcat: concatenate data files
:sdpstep: run a command line tool and save info to process-steps file

Modules:
----------------------

:lcheapo: functions accessing different parts of LCHEAPO 2000 files

Other subdirectories
======================

`lcheapo/_examples/`
------------------------------------------------------------

Should contain example information files and scripts:


CHANGELOG.rst
VERSIONS (changes are to lcfix unless otherwise noted)
 0.2:
  - Handles spliced datafiles with zero-filled gaps:
    + Adds dates to headers in gap
    + Verifies that gap is the right size
      (no time tear afterwards)
  - Handles incomplete datafiles
    + Warns the user that the data do not go as far as the
      directory claims
    + Changes the "directoryEntries" value in the header
0.3:
  - Output starts with block #
  - Output is streamlined for zero- filled gaps
  - Confirm channel #: consecutive and < 4)
  - Changed block# reported to conform to lcdump.py (tell()/512-1)
  - Expanded BUG2 to handle 2 consecutive bad times/channel
    (called BUG #2b)
  - Time tears (MUST BE FIXED) are saved to timeTears.txt
  - Bug #1s repeating every 500 blocks are identified as Bug #3
    (can reduce the text output by a lot!)
0.4:
  - Expanded BUG #2 to handle up to 3 consecutive bad times/channel
    (BUG #2c)
0.45:
  - Add --dryrun option (don't output fixed LCHEAPO file)
  - Add check of non-time header values
0.46:
  - By default, program names output file
0.5:
  - Creates a JSON file with execution information
  - Handles multiple lcheapo input files (should be different fragments of
    same station, still only write one JSON file)
0.51:
  - Names the JSON file based on the input filename (each input file
    gets its own JSON)
  - Add -F (forceTimes) option: to force time tags to be consecutive (only
    use if all time tears are proven wrong)
0.52:
  - The JSON file is ALWAYS named process-steps.json
  - If there is already a process-steps.json, new information is appended
0.6:
  - If multiple files specified, assume all are sections of same instrument
    file, add header from first file to all others, output names include
    timestamp of start of data
  - If there is already a process-steps.json, new information is appended
0.61:
  - Corrected process-steps.json to have steps as list, not dictionary
0.62:
  - Fixed bug with "warnings" variable
0.63:
  - Fixed bug making last directory entry for originally headerless files
0.64:
  - Updated process-steps.json file to match process-steps.schema.json
0.65:
  - Shifts to "force time stamp" for last 3 samples: avoids reading
    beyond EOF
0.66:
  - Added '-d', '-i' and '-o' options to match SDPCHAIN programs
0.67:
  - Recognize '*.header.*' files as header+directory, without data
  - DOESN"T WORK CORRECTLY!! CAT HEADER TO FIRST DATA!!!!

v0.70
------

Last version that only works on Python 2.7
- First version using unittest
- Added sdpchain.py (for process-steps.json files)

v0.71
------

Converted to Python 3.6
- Numerous flake8 fixes
- lcdump argparser replaces optparser

v0.72
-------

Repairs bug in old software version (8.07J) used on 2-channel SPOBSs (SPOBS1) in which the directory
length was marked as 16386 blocks when it is actually 14336

  - 0.72.5: fixed error where a file named {SOMETHING}.fix.lch would be
    overwritten
  - 0.72.6: made lcfix counters a class.  Fixed a crash when
    muxChannel == numChannels (1 more than is possible).  Removed
    handling of zero-filled blocks (gives error now)
  - 0.72.7: lcheader now creates a process-steps.json file
  - 0.72.8: `lcheader` outputs header.lch instead of header.raw.lch. `lcinfo`
            no longer creates a process-step (no processing!). `lcdump` added
            a --from_end option to show blocks at end of file
  - 0.72.9: `lcheader` can be run non-interactively using command-line
            arguments
  - 0.72.10: adds SDPCHAIN protocol to lcfix, lcheader and lccut
  - 0.72.11: modify SDPCHAIN to read process-steps from in_dir and write
    to out_dir.  lcfix now quits if there is already a fix.lch file at
    the destination. 
    
v0.73
-------

Repairs lcfix and lcheader bug in which dirBlock was not recalculated when
dirCount was.
   - post2: readheader() does not quit if unable to read software version
      (now returns status = 0)
  - 0.73.1: Heavy changes to underlying sdpchain module, should make more
            consistent and compatible process-steps.json files

v1.0
-------
Name change from `lcheapo`, to `lcheapo_noobspy` (old `lcheapo` and
`lcheapo_obspy` combined into current `lcheapo`)

todo::
- Make output file simply replace raw.lch (or orig.lch) by fix.lch
    - Maybe reject all other suffixes, except header.lch?

lcheapo/lcheader.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
Create/modify an LCHEAPO data file header
"""
# from __future__ import (absolute_import, division, print_function,
#                         unicode_literals)
# from future.builtins import *  # NOQA @UnusedWildImport

import sys
# import os
import math as m
import argparse
from datetime import datetime, timedelta
from pathlib import Path

from .lcheapo import (LCDiskHeader, LCDirEntry)
from .version import __version__

from . import sdpchain

# ------------------------------------
# Global Variable Declarations
# ------------------------------------
PROGRAM_NAME = "lcheader"
bytes_per_block = 512
dirEntries_per_dirBlock = 16
blocks_per_dirEntry = 14336
samples_per_block = 166
accepted_sample_rates = [31.25, 62.5, 125, 250, 500, 1000, 2000, 4000]
accepted_num_channels = [1, 2, 3, 4]


def main():
    """main function"""
    opts = __getOptions()
    
    params = dict(wake_time=datetime(2017, 1, 1, 5, 0, 0),
                  end_time=datetime(2018, 1, 1, 5, 0, 0),
                  output_filename='generic.header.lch')
    if opts.input_file:
        h = __read_header(Path(opts.in_dir) / opts.input_file)
    else:
        h = __generic_header()
    if opts.description:
        h.description = opts.description
    if opts.sample_rate:
        h.realSampleRate = opts.sample_rate
        h.sampleRate = m.floor(opts.sample_rate)
    if opts.num_channels:
        h.numberOfChannels = opts.num_channels
    if opts.wake_time:
        params['wake_time'] = __get_datetime(None, opts.wake_time)
    if opts.end_time:
        params['end_time'] = __get_datetime(None, opts.end_time)
    if opts.output_file:
        params['output_filename'] = opts.output_file
    h, params = _modify_parameters(h, params, opts)
    if opts.dry_run:
        sys.exit(2)
    with open(Path(opts.out_dir) / params['output_filename'], 'wb') as fp:
        # Pre-blank file
        if opts.noDirectory:
            fp.write(b'\x00' * bytes_per_block * h.dirStart)
        else:
            fp.write(b'\x00' * bytes_per_block * h.dataStart)

        # Write header
        h.seekHeaderPosition(fp)
        h.writeHeader(fp)
        if opts.noDirectory:
            sys.exit(0)

        # Write directory
        d = __prep_dirEntry(h.sampleRate)
        dt = params['wake_time']
        dirCount = 0
        samples_per_dirEntry = samples_per_block * blocks_per_dirEntry / \
            h.numberOfChannels
        dir_timeoffset = timedelta(seconds=samples_per_dirEntry / h.sampleRate)
        blocknum = h.dataStart
        d.seekBlock(fp, h.dirStart)
        while dt < params['end_time']:
            d.blockNumber = blocknum
            d.changeTime(dt)
            d.writeDirEntry(fp)
            dt += dir_timeoffset
            dirCount += 1
            blocknum += blocks_per_dirEntry
        h.dirCount = dirCount
        h.dirBlock = h.dirStart + int(dirCount/dirEntries_per_dirBlock)
        h.seekHeaderPosition(fp)
        h.writeHeader(fp)
        returnCode = 0
        params['wake_time'] = params['wake_time'].isoformat()
        params['end_time'] = params['end_time'].isoformat()

        sdpchain.make_process_steps_file(
            opts.in_dir,
            opts.out_dir,
            'lcheader',
            'Create an LCHEAPO header and directory',
            __version__,
            " ".join(sys.argv),
            datetime.strftime(datetime.utcnow(),  '%Y-%m-%dT%H:%M:%S'),
            returnCode,
            exec_parameters=params)
    sys.exit(returnCode)


def __getOptions():
    """
    Parse user passed options and parameters
    """
    # usageStr = "%prog [-h] [-V]"
    parser = argparse.ArgumentParser(
        description="Create/modify an LCHEAPO data file header",
        epilog="Interactive unless you specify --description, -s, -c, -w, -e "
               "and --output_file")
    parser.add_argument("-V", "--version", default=False, action="store_true",
                        help="Display Program Version")
    parser.add_argument("-n", "--noDirectory", default=False,
                        action="store_true", help="don't create a directory")
    parser.add_argument("--dry_run", default=False, action="store_true",
                        help="don't save to disk")
    parser.add_argument("--no_questions", default=False, action="store_true",
                        help="make a generic header, no questions", )
    parser.add_argument("--input_file",
                        help="input header file to modify")
    parser.add_argument("--description",
                        help="Header Description field (e.g. "
                             "Experiment_OBSType_SN_Station)")
    parser.add_argument("-s", "--sample_rate", type=float, metavar="SAMP_RATE",
                        choices=accepted_sample_rates,
                        help=f"Sample Rate. Allowed = {accepted_sample_rates}")
    parser.add_argument("-c", "--num_channels", type=int, metavar="NCHANS",
                        choices=accepted_num_channels,
                        help=f"Number of channels. Allowed = {accepted_num_channels}")
    parser.add_argument("-w", "--wake_time",  metavar="TIME",
                        help="Start of recorded data (ISO0601 format)")
    parser.add_argument("-e", "--end_time",  metavar="TIME",
                        help="End of recorded data (ISO0601 format)")
    parser.add_argument("--output_file", help="Output filename")
    parser.add_argument("-d", "--directory", dest="base_dir",
                        default='.', help="Base directory for files")
    parser.add_argument("-i", "--input", dest="in_dir", default='.',
                        help="path to input files (abs, or rel to base)")
    parser.add_argument("-o", "--output", dest="out_dir", default='.',
                        help="path for output files (abs, or rel to base)")
    args = parser.parse_args()

    args.in_dir, args.out_dir = sdpchain.setup_paths(args.base_dir,
                                                     args.in_dir,
                                                     args.out_dir)
    args = parser.parse_args()

    # Get the filename (the arguments)
    if args.version:
        print(f"{PROGRAM_NAME} - Version: {__version__}")
        sys.exit(0)
    return args


def _modify_parameters(h, params, opts):
    """Imitate LCHEAPO parameter menu, return values"""
    filename_prefix = params['output_filename'].split('.')[0]
    while not __validate_params(h, params, opts):
        h.description = __get_string("Description (Cruise_InstModel_SN_Site)",
                                     h.description)
        h.realSampleRate = __get_float("Sample Rate", h.realSampleRate,
                                       accepted_sample_rates)
        h.sampleRate = m.floor(h.realSampleRate)
        h.numberOfChannels = __get_int("Number of Channels",
                                       h.numberOfChannels,
                                       accepted_num_channels)

        params['wake_time'] = __get_datetime("Wake Time", params['wake_time'])
        if params['end_time'] < params['wake_time']:
            params['end_time'] = params['wake_time'] + timedelta(days=365)
        params['end_time'] = __get_datetime("End Time", params['end_time'])
        filename_prefix = __get_string("Output Filename Prefix",
                                       filename_prefix)
        params['output_filename'] = filename_prefix + '.header.lch'

    datalen = params['end_time'] - params['wake_time']
    datalen_seconds = datalen.days * 86400 + datalen.seconds + \
        datalen.microseconds / 1.e6
    datalen_blocks = int(datalen_seconds * h.sampleRate / samples_per_block) *\
        h.numberOfChannels
    h.writeBlock = h.dataStart + datalen_blocks

    params['sample_rate'] = h.realSampleRate
    params['number_of_channels'] = h.numberOfChannels
    params['description'] = h.description

    return h, params


def __validate_params(h, params, opts):
    """Validate header parameters"""
    if opts.no_questions:
        return True

    print('**** PARAMETERS ****')
    print(f'         Description: {h.description}')
    print(f'         Sample Rate: {h.realSampleRate:g}')
    print(f'  Number of Channels: {h.numberOfChannels}')
    print(f'           Wake Time: {params["wake_time"].isoformat()}')
    print(f'            End Time: {params["end_time"].isoformat()}')
    print(f'     Output Filename: {params["output_filename"]}')

    if opts.description and opts.sample_rate and opts.num_channels \
            and opts.wake_time and opts.end_time and opts.output_file:
        return True

    resp = input('Is this acceptable? [y/N]: ')
    if not resp:
        return False
    if resp[0].lower() == 'y':
        return True
    return False


def __read_header(filename):
    """Read in header from an LCHEAPO file"""
    h = LCDiskHeader()
    with open(filename, 'r') as fp:
        h.seekHeaderPosition(fp)
        h.readHeader(fp)
    return h


def __generic_header():
    """Create a generic lcheapo header"""

    default_data_startblock = 3586
    default_dir_startblock = 8
    default_data_type = 2
    default_data_type_string = "Uncompressed (24-Bit)"

    h = LCDiskHeader()

    # UNUSED OR STANDARD VALUES
    h.dataType = default_data_type
    h.diskSize = 0    # GB
    h.ramSize = 8     # MB
    h.softwareVersion = '9.08a-OLD'
    h.dirStart = default_dir_startblock
    h.dataStart = default_data_startblock
    # h.dirSize=(h.dataStart-h.dirStart-1)*dirEntries_per_dirBlock
    # Modified to work with lc2ms v1
    h.dirSize = (h.dataStart - h.dirStart) * dirEntries_per_dirBlock - 1
    (h.slowStart, h.slowSize, h.slowBlock, h.slowByte) = (0, 0, 0, 0)
    (h.logStart, h.logSize, h.logBlock, h.logByte) = (0, 0, 0, 0)
    (h.slowDataRate, h.slowStartChannel, h.slowNumberOfChannels) = (0, 0, 0)
    (h.readBlock, h.startChannel, h.diskNumber) = (0, 0, 0)
    (h.numberOfWindows, h.readByte, h.writeByte) = (0, 0, 0)

    # These should be modified after writing directory
    h.dirBlock, h.dirCount = (0, 0)

    # USER-SPECIFIED VALUES
    h.description = 'generic lcheapo header'
    h.sampleRate = 125
    h.numberOfChannels = 4
    h.writeBlock = h.dataStart

    # Additions which cannot be written to file
    h.realSampleRate = h.getRealSampleRate(h.sampleRate)
    h.dataTypeString = default_data_type_string
    return h


def __prep_dirEntry(sample_rate):
    """Prepare lcheapo directory entry object"""
    d = LCDirEntry()
    d.recordLength = 0  # Unused
    d.sampleRate = sample_rate
    d.numBlocks = blocks_per_dirEntry
    d.flag = 0x49
    d.muxChannel = 0
    d.U1 = 0            # Unused
    return d


def __get_string(question, default):
    """Read a string from the command line"""
    resp = input("{} [{}]: ".format(question, default))
    if not resp:
        return default
    return resp


def __get_datetime(question, default):
    """Read a datetime from the command line"""
    if question is not None:
        resp = input("{} [{}]: ".format(question, default.isoformat()))
        if not resp:
            return default
    else:
        resp = default
    try:
        resp = datetime.strptime(resp, '%Y-%m-%dT%H:%M:%S')
    except ValueError:
        try:
            resp = datetime.strptime(resp, '%Y-%m-%dT%H:%M:%S.%f')
        except ValueError:
            print('Invalid input: {}'.format(resp))
            if question is not None:
                resp = __get_datetime(question, default)
            else:
                sys.exit(2)
    return resp


def __get_int(question, default, possibles=False):
    """Read an int from the command line"""
    resp = input("{} [{:d}]: ".format(question, default))
    if not resp:
        return default
    else:
        try:
            resp = int(resp)
        except ValueError:
            print('Invalid input: {}'.format(resp))
            resp = __get_int(question, default, possibles)
        if isinstance(possibles, list):
            if resp not in possibles:
                print('Input must be one of: {}'.format(str(possibles)))
                resp = __get_int(question, default, possibles)
    return resp


def __get_float(question, default, possibles=False):
    """Read a float from the command line"""
    resp = input("{} [{:g}]: ".format(question, default))
    if not resp:
        return default
    else:
        try:
            resp = float(resp)
        except ValueError:
            if isinstance(possibles, list):
                print('Input must be one of: {}'.format(str(possibles)))
            else:
                print('Invalid input: {}'.format(resp))
            resp = __get_float(question, default, possibles)
        if isinstance(possibles, list):
            if resp not in possibles:
                print('Input must be one of: {}'.format(str(possibles)))
                resp = __get_float(question, default, possibles)
    return resp


# ---------------------------------------------------------------------------
# Run 'main' if the script is not imported as a module
# ---------------------------------------------------------------------------
if __name__ == '__main__':
    main()
